### ISSUE TITLE
Potential bug in non-canonical UMP2

### ISSUE THREAD

**Issue by hungpham2017** [pyscf/pyscf#1988]
**Description**: I ran UMP2 and RMP2 on the same _close-shell_ carbon atom using arbitrary orbitals (non-converged SCF) and expected UMP2 and RMP2 return the same energy, but they were not.

**Example:**

```
from pyscf import gto, scf, mp

mol = gto.Mole(atom = [['C', 0,0,0]], basis = 'cc-pvdz')
mf = scf.RHF(mol)
mf.kernel()

# Running non-canonical MP2
mf.mo_coeff[:2:,2:] = 0
mf.converged = False
e_ump2 = mf.to_uhf().MP2().kernel()[0]
e_rmp2 = mf.MP2().kernel()[0]
print("E(UMP2) = ", e_ump2)
print("E(RMP2) = ", e_rmp2)
```

**Result**:
```
E(UMP2) =  0.6143361818889358
E(RMP2) =  0.6127504763660901
```
The energies are different. I found that the [update_amps](https://github.com/pyscf/pyscf/blob/148a6fc671e4ba17f64ef3ff74e5676c8de2cb8b/pyscf/mp/ump2.py#L129) function used in non-canonical UMP2 doesn't update the amplitudes when executed by the [_iterative_kernel](https://github.com/pyscf/pyscf/blob/148a6fc671e4ba17f64ef3ff74e5676c8de2cb8b/pyscf/mp/mp2.py#L80).
In line [157-159](https://github.com/pyscf/pyscf/blob/148a6fc671e4ba17f64ef3ff74e5676c8de2cb8b/pyscf/mp/ump2.py#L157), simply change:
```
    u2aa = eris_ovov.transpose(0,2,1,3) - eris_ovov.transpose(0,2,3,1)
    u2bb = eris_OVOV.transpose(0,2,1,3) - eris_OVOV.transpose(0,2,3,1)
    u2ab = eris_ovOV.transpose(0,2,1,3)
```
to 

```
    u2aa += eris_ovov.transpose(0,2,1,3) - eris_ovov.transpose(0,2,3,1)
    u2bb += eris_OVOV.transpose(0,2,1,3) - eris_OVOV.transpose(0,2,3,1)
    u2ab += eris_ovOV.transpose(0,2,1,3)
```
ran the same code again, the energies now agree:
```
E(UMP2) = 0.6127504763657433
E(RMP2) = 0.6127504763660907
```

I would appreciate if someone can have a look on this. Thanks

**Comment by tberkel** [comment]
Thanks @hungpham2017 this looks like a bug to me. Please submit a PR.

Related, in our group we recently realized that the iterative MP2 code logic seems questionable (tagging @hongzhouye). An unconverged SCF shouldn't necessarily trigger an iterative MP2. An unconverged SCF simply defines a non-HF determinant that should have a singles contribution to the one-shot MP energy. The more conventional use of iterative MP2 is when a non-canonical set of HF orbitals is used. Checking for this would require a "canonical" check (e.g., if SCF converged and Fock is diagonal in the basis of `mo_coeff`). Am I missing something?

**Comment by hongzhouye** [comment]
@hungpham2017 not directly related to the bug: what's your thought of doing semi-canonicalization instead of using the iterative solver (may also want to add a non-Brillouin single contribution for the non-HF reference)?

**Comment by hungpham2017** [comment]
@tberkel: I agree that a canonical check would be more logical in deciding when to execute the iterative solver. I'm curious if both converged and unconverged SCF with canonical orbitals, i.e., those that diagonalize the Fock matrix, could be categorized under the canonical case. Would everything else then fall under non-canonical?

Currently in the MP2 kernel at [L600](https://github.com/pyscf/pyscf/blob/master/pyscf/mp/mp2.py#L600), 
```
        if self._scf.converged:
            self.e_corr, self.t2 = self.init_amps(mo_energy, mo_coeff, eris, with_t2)
        else:
            self.converged, self.e_corr, self.t2 = _iterative_kernel(self, eris)
```

maybe it can be something like:

```
        if canonical:
            self.e_corr, self.t2 = self.init_amps(mo_energy, mo_coeff, eris, with_t2)
        else:
            self.converged, self.e_corr, self.t2 = _iterative_kernel(self, eris)
```

@hongzhouye: I believe that a non-iterative solver utilizing semi-canonicalized orbitals would be more economically advantageous. Regarding accuracy, do non-iterative and iterative solvers yield the same energy? Or should we allow users to decide which solver to use?

something like this:

```
        if canonical:
            self.e_corr, self.t2 = self.init_amps(mo_energy, mo_coeff, eris, with_t2)
        elif semicanonical:
            self.e_corr, self.t2 = _non_iterative_kernel(self, eris)
        else:
            self.converged, self.e_corr, self.t2 = _iterative_kernel(self, eris)
```

**Comment by tberkel** [comment]
Yes, non-iterative and iterative solvers should yield the same energy as long as they are done for the same reference wavefunction (whether canonical or not). Iterative solvers are most common when a non-canonical representation of the orbitals is intentional, e.g., with localized orbitals for use in local correlation methods, wherein approximations will be introduced.

**Comment by sunqm** [comment]
This issue is fixed by PR #1991
